d2/src/test/java/com/linkedin/d2/balancer/clusterfailout/FailedoutClusterManagerTest.java

```diff
@@ -0,0 +1,129 @@
+/*
+   Copyright (c) 2022 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.d2.balancer.clusterfailout;
+
+import com.linkedin.d2.balancer.LoadBalancerState;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import static org.mockito.Mockito.*;
+import static org.testng.Assert.*;
+
+public class FailedoutClusterManagerTest
```

Comment:
I'm writing an integration test on D2 that simulates a real server setup (by announcing to local ZK instance) and spins up clients that send requests to the servers in parallel. I think we should be able to utilize this test infra pieces and also have an e2e integration test for fail out. We can discuss more offline